build(docs-infra): associate providedIn injectables with their NgModule

petebacondarwin opened the pull request:
Such injectables were not appearing in the providers lists of their NgModule.

The first commit updates the doc-gen to support associating these automatically.

Further, it also allows developers to mark other injectables that are provided
in an NgModule with a reference to the NgModule where they are provided.

The commit also does a refactoring of the `processNgModuleDocs` dgeni
processor code, to make it easier to maintain.

The second commit updates all classes that are provided by NgModules.
This allows the doc-gen to add links back to the NgModule from
the class's documentation page.

Fixes #41203

petebacondarwin committed:
build(docs-infra): associate `providedIn` injectables with their NgModule

Such injectables were not appearing in the providers lists of their NgModule.
This commit updates the doc-gen to support associating these automatically.
Further, it also allows developers to mark other injectables that are provided
in an NgModule with a reference to the NgModule where they are provided.

The commit also does a refactoring of the `processNgModuleDocs` dgeni
processor code, to make it easier to maintain.

Fixes #41203

mary-poppins commented:
You can preview fcb50f7 at https://pr41960-fcb50f7.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-api-package/processors/processNgModuleDocs.js:
```diff
@@ -5,77 +5,196 @@ module.exports = function processNgModuleDocs(getDocFromAlias, createDocMessage,
+ * Convert all the values of properties on the `obj` to arrays, if not already.
+ */
+function convertAllPropertiesToArrays(obj) {
+  for (const key of Object.keys(obj)) {
+    const value = obj[key];
```
Nit:

```suggestion
  for (const [key, value] of Object.entries(obj)) {
```

gkalpak commented on aio/tools/transforms/angular-api-package/processors/processNgModuleDocs.js:
```diff
@@ -5,77 +5,196 @@ module.exports = function processNgModuleDocs(getDocFromAlias, createDocMessage,
-        errors.forEach(error => log.error(error));
-        throw new Error('Failed to process NgModule relationships.');
+    /**
+     * Associate the given `injectableDoc` with an NgModule doc if is provided on an `NgModule`.
```
```suggestion
     * Associate the given `injectableDoc` with an NgModule doc if it is provided on an `NgModule`.
```

gkalpak reviewed (commented):
A couple of questions:

- Could we somehow auto-detect `NgModule`s that provide each injectable? Otherwise, the process feels very error-prone (and prone to becoming stale) 😟
- I think we should detect and document when injectables are provided in `'root'`/`'platform'` (either directly via `providedIn: '...'` or via the [de-sugared injection form](https://github.com/angular/angular/blob/fcb50f77610b9478050e58f0d0448a2a7790400d/packages/common/src/viewport_scroller.ts#L25-L29) :scream:). Otherwise, the docs seem to confusingly imply that a certain injectable would only be provided by a testing module or something.

<sub>
(And no, I didn't notice that `@ngModule` sometimes appears right above `@publicApi` and sometimes right below :see_no_evil: :see_no_evil: :see_no_evil:)
</sub>

petebacondarwin commented:
> Could we somehow auto-detect NgModules that provide each injectable? Otherwise, the process feels very error-prone (and prone to becoming stale) 😟

I thought about this. It is a bit tricky but might be possible. We could parse the provider strings that look like:

```
"{ provide: Xxx, ... }"
```

and extract the reference to the thing being provided (e.g. `Xxx`).

I was led to the current solution, by the fact that we already require directives and pipes to have one or more `@ngModule` tags that link them to the place they are exported.

Let me have a further play with that idea.

petebacondarwin commented:
> I think we should detect and document when injectables are provided in 'root'/'platform' (either directly via providedIn: '...' or via the de-sugared injection form 😱). Otherwise, the docs seem to confusingly imply that a certain injectable would only be provided by a testing module or something.

I think we could add something to the `NgModules` section of such injectables.
I'm not sure how easy it will be to identify those de-sugared ones.

petebacondarwin committed:
fixup! build(docs-infra): associate `providedIn` injectables with their NgModule

petebacondarwin committed:
fixup! build(docs-infra): associate `providedIn` injectables with their NgModule

petebacondarwin committed:
refactor(docs-infra): remove unused CSS rule

The `.ngmodule-iist` is no longer used so can be removed.

petebacondarwin committed:
fixup! build(docs-infra): associate `providedIn` injectables with their NgModule

mary-poppins commented:
You can preview 396f142 at https://pr41960-396f142.ngbuilds.io/.

petebacondarwin commented:
Now providedIn `root` and `platform` get visibility on the injectable's page.

* https://pr41960-396f142.ngbuilds.io/api/common/Location#provided-in
* https://pr41960-396f142.ngbuilds.io/api/common/PlatformLocation#provided-in

I also renamed the sections from `NgModules` to `Provided in` (for injectables) and `Exported from` (for directives and pipes).

mary-poppins commented:
You can preview fd142b0 at https://pr41960-fd142b0.ngbuilds.io/.

petebacondarwin committed:
refactor(core): remove unnecessary cast

The `as any` is no longer needed since `providedIn` can accept `'any'`.

petebacondarwin committed:
build(docs-infra): add NgModules references to provided injectables

Parse the `providers` property of each NgModule doc and add this doc to
the `ngModules` property of each provided injectable.

petebacondarwin commented:
I have no refactored so that it automatically wires up injectables that are found in the `{ providedIn: Xxxx, ... }` blocks of each NgModule, so I have removed the commit that added all the `@ngModule` JSDOC tags.

petebacondarwin committed:
fixup! build(docs-infra): add NgModules references to provided injectables

mary-poppins commented:
You can preview 927c258 at https://pr41960-927c258.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-api-package/processors/processNgModuleDocs.js:
```diff
@@ -5,77 +5,245 @@ module.exports = function processNgModuleDocs(getDocFromAlias, createDocMessage,
+        errors.forEach(error => log.error(error));
+        throw new Error('Failed to process NgModule relationships.');
+      }
 
```
```suggestion
```

gkalpak commented on aio/tools/transforms/angular-api-package/processors/processNgModuleDocs.js:
```diff
@@ -5,77 +5,245 @@ module.exports = function processNgModuleDocs(getDocFromAlias, createDocMessage,
+        return;
+      }
+      if (isWrappedInQuotes(providedIn)) {
+        // `providedIn` is wrapped in quotes, so it will be one of `'root'` or `'platform'` and
```
It could also be `'any'`, right?

petebacondarwin replied:
Right!

gkalpak commented on aio/tools/transforms/angular-api-package/processors/processNgModuleDocs.js:
```diff
@@ -5,77 +5,245 @@ module.exports = function processNgModuleDocs(getDocFromAlias, createDocMessage,
+          providers.push(`{ provide: ${provider.name}, useClass: ${provider.name} }`);
+        }
+      }
+      if (providers.length > 0) {
+        ngModuleDoc.providers = providers;
+      }
+
     }
```
```suggestion

      if (providers.length > 0) {
        ngModuleDoc.providers = providers;
      }
    }
```

(???)

petebacondarwin replied:
😕

gkalpak reviewed (approved):
A couple of minor comments - otherwise lgtm :shipit: 
Amazing work :100:

AndrewKushnir commented:
[Presubmit](http://test/OCL:372725829:BASE:372725828:1620493031559:58fc00cc).
